pkg/controller/controller_test.go

```diff
@@ -2039,6 +2072,8 @@ func assertServiceInstanceErrorBeforeRequest(t *testing.T, obj runtime.Object, r
 	assertServiceInstanceCurrentOperationClear(t, obj)
 	assertServiceInstanceOperationStartTimeSet(t, obj, false)
 	assertServiceInstanceReconciledGeneration(t, obj, originalInstance.Status.ReconciledGeneration)
+	assertServiceInstanceObservedGeneration(t, obj, originalInstance.Status.ObservedGeneration)
```

Comment:
I would expect that the controller would set `ObservedGeneration` even if there was an error prior to sending the request to the broker. For example, if there were an error getting the service class, the controller sets the condition for the instance to `Ready:False`. I would expect that the controller would also set the `ObservedGeneration` to the current generation.

Reply:
Yeah good point, but as I mentioned above it breaks the condition to check whether the instance has been reconciled or not. Probably keeping both `ObservedGeneration` and `ReconciledGeneration` will help there.

Reply:
@staebler I checked the `controller_instance.go` - currently we invoke `recordStartOfServiceInstanceOperation()` method only after we have checked the service class/plan and prepared an OSB request.
Do you suggest that we need to move this invocation higher (right after checking if we need to process at all)?
@kibbles-n-bytes do you agree?

Reply:
The problem here is that `recordStartOfServiceInstanceOperation()` method sets the `instance.Status.InProgressProperties` field that requires service class/plan.
Thus, we should either split `recordStartOfServiceInstanceOperation()` into two methods (the first updating `ObservedGeneration` and the second populating the operation details), or leave it as it is.

Reply:
If we wait until after the class/plan are checked, then the `ObservedGeneration` is inconsistent in the presence of an errors with the check. Let's say that there was an error getting the service class. The controller will change the instance to `Ready:False`. If the controller does not also set the `ObservedGeneration`, then it will appear as though there was an error with an earlier generation as opposed to there being an error reconciling the current generation.

Reply:
> Yeah good point, but as I mentioned above it breaks the condition to check whether the instance has been reconciled or not.

@nilebox I don't understand this. Why does it break the condition? I would argue that it breaks the condition if we do *not* set the observed generation.